leases/tasks/send_invoices.py

```diff
@@ -0,0 +1,231 @@
+from customers.models import CustomerProfile
+from customers.services import ProfileService
+from leases.enums import LeaseStatus
+from leases.models import BerthLease, WinterStorageLease
+from leases.utils import calculate_season_end_date, calculate_season_start_date
+from payments.enums import OrderStatus
+from payments.models import BerthPriceGroup, BerthProduct, Order, WinterStorageProduct
+from payments.utils import approve_order
+
+User = get_user_model()
+
+BERTH_LEASES = "berth_leases"
+WINTER_LEASES = "winter_leases"
+
+
+def fail_lease(
+    lease: Union[BerthLease, WinterStorageLease],
+    message: str,
+    failed_leases: List[Dict[UUID, str]],
+) -> None:
+    """Set a lease to ERROR status and append it to the failed_lease list"""
+    failed_leases.append({lease.id: message})
+    lease.status = LeaseStatus.ERROR
+    lease.save(update_fields=["status"])
+
+
+def fail_order(
+    order: Order, message: str, failed_orders: List[Dict[UUID, str]]
+) -> None:
+    """Set an order to ERROR status and append it to the failed_order list"""
+    failed_orders.append({order.id: message})
+    order.set_status(OrderStatus.ERROR, f"Lease renewing failed: {message}")
+
+
+def get_product(
+    lease: Union[BerthLease, WinterStorageLease]
+) -> Union[BerthProduct, WinterStorageProduct]:
+    """Get the product associated to the lease"""
+
+    if isinstance(lease, BerthLease):
+        return get_berth_product(lease)
+
+    return get_winter_storage_product(lease)
+
+
+def get_berth_product(lease: BerthLease) -> BerthProduct:
+    # Many pre-imported leases don't have a boat associated, so we try to get the
+    # width based on the berth_type.
+    width = lease.boat.width if lease.boat else lease.berth.berth_type.width
+
+    products_for_pricegroup = BerthProduct.objects.filter(
+        price_group=BerthPriceGroup.objects.get_for_width(width),
+    )
+
+    # Get the product for the associated Harbor
+    if harbor_product := products_for_pricegroup.filter(
+        harbor=lease.berth.pier.harbor
+    ).first():
+        product = harbor_product
+
+    # If there's no Harbor-specific product, try with the "default" product
+    else:
+        product = products_for_pricegroup.filter(harbor__isnull=True).first()
+
+    return product
+
+
+def get_winter_storage_product(lease: WinterStorageLease) -> WinterStorageProduct:
+    raise NotImplementedError
+
+
+def get_valid_leases(season_start: date, lease_type: str) -> QuerySet:
+    if lease_type == BERTH_LEASES:
+        return get_valid_berth_leases(season_start)
+    return get_valid_winter_storage_leases(season_start)
+
+
+def get_valid_winter_storage_leases(season_start: date) -> QuerySet:
+    raise NotImplementedError
+
+
+def get_valid_berth_leases(season_start: date) -> QuerySet:
+    """
+    Get the leases that were active last year
+    If today is:
+      (1) before season: leases from last year
+      (2) during or after season: leases from this year
```

Comment:
It's possible to renew leases after the season has ended? It might be clearer for the admins if this action is only performed on currently active leases? I feel the already expired ones should be manually renewed.

Reply:
I'm not sure about the business requirements for this, I'll bring it up on Slack. 👍

Reply:
Rsolved over slack:

> Lets go on with the case 2 that doesn’t limit the invoicing timetable and covers all the leases from previous season.
>
> Recurring invoicing will most probably always happen straight after the season before the year changes, but I don’t see any reason to limit the the invoicing from the system side.